File: my-folder/0260-single-number-iii/solution.java
class Solution {
    public int[] singleNumber(int[] nums) {
       int xorResult = 0;
        for (int num : nums) {
            xorResult ^= num;
        }

        int rightmostSetBit = xorResult & -xorResult;

        int group1 = 0;
        int group2 = 0;
        for (int num : nums) {
            if ((num & rightmostSetBit) != 0) {
                group1 ^= num;
            } else {
                group2 ^= num;
            }
        }

        return new int[]{group1, group2}; 
    }
}

File: my-folder/0260-single-number-iii/solution.py
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        xorResult = 0
        for num in nums:
            xorResult ^= num
        
        xorResult &= -xorResult

        res = [0, 0]

        for num in nums:
            if num & xorResult == 0:
                res[0] ^= num
            else:
                res[1] ^= num
        
        return res
        
